Add Azure integrations overview (#2285)
dotnet/docs-aspire: docs/azure/integrations-overview.md

- What happens when these resources are deployed?
- How does these show up in the manifest?
- What does azd do here?

-->

# .NET Aspire Azure integrations overview

[Azure](/azure) is the most popular cloud platform for building and deploying [.NET applications](/dotnet/azure). The [Azure SDK for .NET](/dotnet/azure/sdk/azure-sdk-for-dotnet) allows for easy management and use of Azure services. .NET Aspire provides a set of integrations with Azure services. This article details some common aspects of all Azure integrations in .NET Aspire and aims to help you understand how to use them.

## Add Azure resources

All .NET Aspire Azure hosting integrations expose Azure resources and by convention are added using `AddAzure*` APIs. When you add these resources to your .NET Aspire app host, they represent an Azure service. The `AddAzure*` API returns an <xref:Aspire.Hosting.ApplicationModel.IResourceBuilder`1> where `T` is the type of Azure resource. These `IResourceBuilder<T>` (builder) interfaces provide a fluent API that allows you to configure the underlying Azure resource within the [app model](xref:dotnet/aspire/app-host#terminology).

### Typical developer experience

> When your .NET Aspire app host contains Azure resources, and you run it locally (typical developer <kbd>F5</kbd> experience), the Azure resources are provisioned in your Azure subscription. However, they're not yet deployed. Instead, they're running locally in the context of your app host.

Reviewer: ```suggestion
When your .NET Aspire app host contains Azure resources, and you run it locally (typical developer <kbd>F5</kbd> or `dotnet run` experience), the Azure resources are provisioned in your Azure subscription. However, they're not yet deployed. Instead, they're running locally in the context of your app host.
```

> When your .NET Aspire app host contains Azure resources, and you run it locally (typical developer <kbd>F5</kbd> experience), the Azure resources are provisioned in your Azure subscription. However, they're not yet deployed. Instead, they're running locally in the context of your app host.

Reviewer: > However, they're not yet deployed. However, they're not yet deployed.

This isn't correct. They are deployed and running in Azure. They are deployed to a subscription and resource group specified for your local app development.

| Hosting integration | Details |
|--|--|
| Azure Cache for Redis | Call <xref:Aspire.Hosting.AzureRedisExtensions.RunAsContainer*?displayProperty=nameWithType> on the `IResourceBuilder<AzureRedisCacheResource>` to configure it to run locally in a container, based on the `docker.io/library/redis` image. |
| Azure PostgreSQL Flexible Server | Call <xref:Aspire.Hosting.AzurePostgresExtensions.RunAsContainer*?displayProperty=nameWithType> on the `IResourceBuilder<AzurePostgresFlexibleServerResource>` to configure it to run locally in a container, based on the `docker.io/library/postgres` image. |
| Azure SQL Server | Call <xref:Aspire.Hosting.AzureSqlExtensions.RunAsContainer*?displayProperty=nameWithType> on the `IResourceBuilder<AzureSqlServerResource>` to configure it to run locally in a container, based on the `mcr.microsoft.com/mssql/server` image. |

> [!NOTE]
> Like emulators, calling `RunAsContainer` on an Azure resource builder has no affect on the [publishing manifest](../deployment/manifest-format.md). When you publish your app, the generated Bicep file will reflect the actual Azure service, not the local container.

> ### Common APIs and patterns

Author: I renamed the section heading, and it's an h3 which is a continuation of the **Add Azure resources** section, if that makes sense?

<!-- TODO: Add link from here to the Azure docs when they're written. -->

While it's possible to provision Azure resources manually, .NET Aspire simplifies the process by providing a set of APIs to express Azure resources. These APIs are available as extension methods in .NET Aspire Azure hosting libraries, extending the <xref:Aspire.Hosting.IDistributedApplicationBuilder> interface. When you add Azure resources to your app host, they add the appropriate provisioning functionality implicitly. In other words, you don't need to call any provisioning APIs directly.

Since .NET Aspire models Azure resources within Azure hosting integrations, the Azure SDK is used to provision these resources. The Azure SDK generates Bicep files that define the Azure resources you need. If you're unfamiliar with Bicep, it's a domain-specific language used to describe and provision Azure resources. The generated Bicep files are output alongside the manifest file when you publish your app.

### Azure provisioning customization

> All .NET Aspire Azure hosting integrations expose various Azure resources, and they're all subclasses of the <xref:Aspire.Hosting.Azure.AzureProvisioningResource> type—which itself inherits the <xref:Aspire.Hosting.Azure.AzureBicepResource>. This enables extensions that are generically type-constrained to this type, allowing for a fluent API to customize the infrastructure to your liking. While the Azure SDK relies on defaults, you're free to influence the generated Bicep using these APIs.

Author: Fixed

#### Configure infrastructure

Regardless of the Azure resource you're working with, to configure its underlying infrastructure, you chain a call to the <xref:Aspire.Hosting.AzureProvisioningResourceExtensions.ConfigureInfrastructure*> extension method. This method allows you to customize the infrastructure of the Azure resource by passing a `configure` delegate of type `Action<AzureResourceInfrastructure>`. The <xref:Aspire.Hosting.Azure.AzureResourceInfrastructure> type is a subclass of the <xref:Azure.Provisioning.Infrastructure?displayProperty=fullName>. This type exposes a massive API surface area for configuring the underlying infrastructure of the Azure resource.

Each `configure` action:

- Calls the <xref:Azure.Provisioning.Infrastructure.GetProvisionableResources> API to get the provisionable resources.

> - Filters the provisionable resources based on target resource type to customize.

Reviewer: This is worded kind of odd. You don't actually need to call GetProvisionableResources. You can just add more, if that's what you want/need.

Maybe it could be just:

"A `configure` action may:"

Reviewer: Or maybe even just delete this part as we explain line-by-line below the code.

Author: I removed it.

- Applies the desired configuration to the resource.

Consider the following example:

:::code language="csharp" source="../snippets/azure/AppHost/Program.ConfigureInfrastructure.cs" id="infra":::

The preceding code:

- Adds a parameter named `storage-sku`.
- Adds Azure Storage with the <xref:Aspire.Hosting.AzureStorageExtensions.AddAzureStorage*> API named `storage`.
- Chains a call to `ConfigureInfrastructure` to customize the Azure Storage infrastructure:
  - Gets the provisionable resources.
  - Filters to a single <xref:Azure.Provisioning.Storage.StorageAccount>.
  - Assigns the `storage-sku` parameter to the <xref:Azure.Provisioning.Storage.StorageAccount.Sku?displayProperty=nameWithType> property:
    - A new instance of the <xref:Azure.Provisioning.Storage.StorageSku> has its `Name` property assigned from the result of the <xref:Aspire.Hosting.AzureProvisioningResourceExtensions.AsProvisioningParameter*> API.

> This exemplifies flowing an [external parameter](../fundamentals/external-parameters.md) into the Azure Storage infrastructure, resulting in the generated Bicep file reflecting the desired configuration.

Reviewer: I'm not sure if we already have it, but it might be nice to explain where the parameter value comes from - or link off to parameters.

Author: That's what the external parameter link is for, it's detailed in there.

- [📦 Azure.Provisioning.OperationalInsights](https://www.nuget.org/packages/Azure.Provisioning.OperationalInsights)
- [📦 Azure.Provisioning.PostgreSql](https://www.nuget.org/packages/Azure.Provisioning.PostgreSql)
- [📦 Azure.Provisioning.Redis](https://www.nuget.org/packages/Azure.Provisioning.Redis)
- [📦 Azure.Provisioning.Search](https://www.nuget.org/packages/Azure.Provisioning.Search)
- [📦 Azure.Provisioning.ServiceBus](https://www.nuget.org/packages/Azure.Provisioning.ServiceBus)
- [📦 Azure.Provisioning.SignalR](https://www.nuget.org/packages/Azure.Provisioning.SignalR)
- [📦 Azure.Provisioning.Sql](https://www.nuget.org/packages/Azure.Provisioning.Sql)
- [📦 Azure.Provisioning.Storage](https://www.nuget.org/packages/Azure.Provisioning.Storage)
- [📦 Azure.Provisioning.WebPubSub](https://www.nuget.org/packages/Azure.Provisioning.WebPubSub)
- [📦 Azure.Provisioning](https://www.nuget.org/packages/Azure.Provisioning)

> [!TIP]
> You don't need to install these packages manually in your app host projects, as they're transitive dependencies of the corresponding .NET Aspire Azure hosting integrations your app host references.

## Publishing

> When you publish your app, the Azure provisioning generated Bicep is used by the Azure Developer CLI to create the Azure resources in your Azure subscription. The Azure Developer CLI is a command-line tool that provides a set of commands to manage Azure resources. For more information, see [Azure Developer CLI](/azure/developer/azure-developer-cli).

Reviewer: just an FYI - but the bicep is used both at local provisioning and pubilshing/deployment time.

- What happens when these resources are deployed?
- How does these show up in the manifest?
- What does azd do here?

-->

# .NET Aspire Azure integrations overview

[Azure](/azure) is the most popular cloud platform for building and deploying [.NET applications](/dotnet/azure). The [Azure SDK for .NET](/dotnet/azure/sdk/azure-sdk-for-dotnet) allows for easy management and use of Azure services. .NET Aspire provides a set of integrations with Azure services. This article details some common aspects of all Azure integrations in .NET Aspire and aims to help you understand how to use them.

## Add Azure resources

All .NET Aspire Azure hosting integrations expose Azure resources and by convention are added using `AddAzure*` APIs. When you add these resources to your .NET Aspire app host, they represent an Azure service. The `AddAzure*` API returns an <xref:Aspire.Hosting.ApplicationModel.IResourceBuilder`1> where `T` is the type of Azure resource. These `IResourceBuilder<T>` (builder) interfaces provide a fluent API that allows you to configure the underlying Azure resource within the [app model](xref:dotnet/aspire/app-host#terminology).

### Typical developer experience

> When your .NET Aspire app host contains Azure resources, and you run it locally (typical developer <kbd>F5</kbd> or `dotnet run` experience), the [Azure resources are provisioned](local-provisioning.md) in your Azure subscription. However, they're not yet deployed. Instead, they're running locally in the context of your app host.

Author: ```suggestion
When your .NET Aspire app host contains Azure resources, and you run it locally (typical developer <kbd>F5</kbd> or `dotnet run` experience), the [Azure resources are provisioned](local-provisioning.md) in your Azure subscription. This allows you as the developer to debug against them locally in the context of your app host.
```

Reviewer: 👍 Thanks!

---
title: .NET Aspire Azure integrations overview
description: Overview of the Azure integrations available in the .NET Aspire.
ms.date: 12/11/2024
uid: azure-integrations-overview
---

# .NET Aspire Azure integrations overview

[Azure](/azure) is the most popular cloud platform for building and deploying [.NET applications](/dotnet/azure).  The Azure SDK for .NET allows for easy management and use of Azure services. .NET Aspire provides a set of integrations with Azure services. This article details some common aspects of all Azure integrations in .NET Aspire and aims to help you understand how to use them.

> ## Provisioning

Reviewer 2: This term is overloaded. I would be more descriptive with what these libraries are for. The industry standard term for these libraries are a "CDK" (cloud development kit). I would also use Infrastructure as code somewhere in here.

Author: I was told I shouldn't use CDK, since this is different than the industries usage of CDK.

Reviewer: @tg-msft mentioned to me in the past to not refer to Azure.Provisioning as the "CDK".

Reviewer 2: Lets not conflate locaal provisioning with the provisioning libraries then. We need a better title.

Author: Added a section specifically to address this

While it's possible to provision Azure resources manually, .NET Aspire simplifies the process by providing a set of APIs to express Azure resources. These APIs are available as extension methods in .NET Aspire Azure hosting libraries, extending the <xref:Aspire.Hosting.IDistributedApplicationBuilder> interface. When you add Azure resources to your app host, they'll add the appropriate provisioning functionality implicitly. In other words, you don't need to call any provisioning APIs directly.

> The **Azure SDK for .NET**, provides the [📦 Azure.Provisioning](https://www.nuget.org/packages/Azure.Provisioning) NuGet package and a suite of [Azure provisioning packages](#azure-provisioning-packages). These Azure provisioning libraries make it easy to declaratively specify Azure infrastructure natively in .NET.

Reviewer 2: I would start this paragraph with this.

Since .NET Aspire models Azure resources within Azure hosting integrations, the Azure SDK is used to provision these resources. The Azure SDK generates Bicep files that define the Azure resources you need. If you're unfamiliar with Bicep, it's a domain-specific language used to describe and provision Azure resources. The generated Bicep files are output alongside the manifest file when you publish your app. For more information, see [Azure provisioning Bicep](#azure-provisioning-bicep).

> ### Azure provisioning packages

Reviewer 2: I don't think we need this list so high in the overview. I think its fine to put it near the bottom with a link to the azure service as its not always obvious what it is for.

- [📦 Azure.Provisioning.SignalR](https://www.nuget.org/packages/Azure.Provisioning.SignalR)
- [📦 Azure.Provisioning.Sql](https://www.nuget.org/packages/Azure.Provisioning.Sql)
- [📦 Azure.Provisioning.Storage](https://www.nuget.org/packages/Azure.Provisioning.Storage)
- [📦 Azure.Provisioning.WebPubSub](https://www.nuget.org/packages/Azure.Provisioning.WebPubSub)
- [📦 Azure.Provisioning](https://www.nuget.org/packages/Azure.Provisioning)

> [!TIP]
> You don't need to install these packages manually in your app host projects, as they're transitive dependencies of the corresponding .NET Aspire Azure hosting integrations your app host references.

### Azure provisioning defaults

The Azure provisioning libraries provide default values for Azure resources. These defaults are only a starting point and aren't intended for all use cases. You can customize these defaults in multiple ways to meet your specific requirements. For more information, see [Azure provisioning customization](#azure-provisioning-customization) and [Use custom Bicep templates](../deployment/azure/custom-bicep-templates.md).

> #### Azure role assignment defaults

Reviewer 2: I think we should delete this.

Author: I thought we wanted to document the default role assignments?

Reviewer 2: In each integration not in the overview

Reviewer: Why should we delete it? It's it valuable for devs to know what role assignments are used?

Note that it will need to be updated when we redo the role assignments stuff in the next release.

Reviewer 2: We don't need detailed information about each integration on the overview page. The point of this page is to explain all of the *shared* ideas and concepts for azure integrations, not a place to roll up all of the details of each one.

Author: Addressed

> #### Azure App Configuration defaults

Author: Addressed
